backend/src/Edna/Upload/Service.hs

```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+  fromMaybe (error $ "added compound not found: " <> compoundName)  <$>
+    compoundNameToId compoundName
+
+insertExperimentFile ::
+  Word32 -> Word32 -> FileMetadata -> Text -> Text -> LByteString -> Edna Word32
+insertExperimentFile projId methodId meta descr fileName blob = do
+  unSerial . expectOneInsertion . map efExperimentFileId <$> runInsertReturningList'
+    (insert (esExperimentFile ednaSchema) $ insertExpressions [experimentFileRec])
+  where
+    experimentFileRec :: ExperimentFileT (QExpr Postgres s)
+    experimentFileRec = ExperimentFileRec
+      { efExperimentFileId = default_
+      , efProjectId = val_ projId
+      , efMethodologyId = val_ (Just methodId)
+      , efUploadDate = default_
+      , efMeta = val_ (PgJSON meta)
+      , efDescription = val_ descr
+      , efName = val_ fileName
+      , efContents = val_ blob
+      }
+
+insertExperiment :: Word32 -> Word32 -> Word32 -> [Measurement] -> Edna ()
+insertExperiment experimentFileId compoundId targetId measurements = do
+  expId <-
+    unSerial . expectOneInsertion . map eExperimentId <$> runInsertReturningList'
+    (insert (esExperiment ednaSchema) $ insertExpressions [experimentRec])
+  subExpId <-
+    unSerial . expectOneInsertion . map seSubExperimentId <$> runInsertReturningList'
+    (insert (esSubExperiment ednaSchema) $ insertExpressions [subExperimentRec expId])
+  forM_ measurements $ \measurement -> do
+    let
+      measurementRec :: MeasurementT (QExpr Postgres s)
+      measurementRec = MeasurementRec
+        { mMeasurementId = default_
+        , mExperimentId = val_ expId
+        , mConcentration = val_ (EReader.mConcentration measurement)
+        , mSignal = val_ (EReader.mSignal measurement)
+        , mIsOutlier = val_ (EReader.mIsOutlier measurement)
+        }
+    measurementId <-
+      unSerial . expectOneInsertion . map mMeasurementId <$> runInsertReturningList'
+      (insert (esMeasurement ednaSchema) $ insertExpressions [measurementRec])
+    when (EReader.mIsOutlier measurement) $
+      runInsert' $ insert (esRemovedMeasurements ednaSchema) $ insertValues
+      [RemovedMeasurementsRec subExpId measurementId]
+  where
+    experimentRec :: ExperimentT (QExpr Postgres s)
+    experimentRec = ExperimentRec
+      { eExperimentId = default_
+      , eExperimentFileId = val_ experimentFileId
+      , eCompoundId = val_ compoundId
+      , eTargetId = val_ targetId
+      }
+
+    subExperimentRec :: Word32 -> SubExperimentT (QExpr Postgres s)
+    subExperimentRec experimentId = SubExperimentRec
+      { seSubExperimentId = default_
+      , seAnalysisMethodId = val_ theOnlyAnalysisMethodId
+      , seExperimentId = val_ experimentId
+      , seIsSuspicious = val_ False
+      , seResult = val_ (PgJSON 10)  -- stub value, will be computed later
+      }
+
+expectOneInsertion :: (HasCallStack, Show x) => [x] -> x
+expectOneInsertion = \case
+  [x] -> x
+  xs -> error $ "Expected to insert 1 item, but inserted: " <> show xs
```

Comment:
Why not `specific` errors?

Reply:
What is `specific`? Anyway, this function will likely be removed because you say there is `runInsertReturningOne`.

Reply:
So, create smth like `data DBError = `. There are also other plain errors in this module

Reply:
Because this is a programmer mistake which should never happen. I generally follow [these guidelines](https://github.com/input-output-hk/cardano-sl/blob/2a5b5769799fd220e2ff6a6b09a1832d8e6f06c0/docs/exceptions.md). `data DBError` is useful when there is a chance that this error will be handled somehow because you can turn it into some other format (e. g. HTTP response) or even translate to another language. But this error is a programmer mistake which we don't need to handle specially.